packages/test/test-end-to-end-tests/src/test/batching.spec.ts

```diff
@@ -531,3 +532,116 @@ describeFullCompat("Batching", (getTestObjectProvider) => {
+        dataObject1.context.containerRuntime.setFlushMode(FlushMode.Immediate);
+        dataObject1map1.set("key5", "value5");
+        dataObject1map2.set("key6", "value6");
+
+        // Send couple of ops in TurnBased FlushMode. These ops should be batched together.
+        dataObject1.context.containerRuntime.setFlushMode(FlushMode.TurnBased);
+        dataObject1map1.set("key3", "value3");
+        dataObject1map2.set("key4", "value4");
+        await yieldJSTurn();
+
+        // Wait for the ops to get processed by both the containers.
+        await provider.ensureSynchronized();
+
+        assert.equal(
+            dataObject1BatchMessages.length, 8, "Incorrect number of messages received on local client");
+
+        // The first couple of ops in Immediate mode should have been sent individually without batch metadata.
+        verifyBatchMetadata(dataObject1BatchMessages.slice(0, 1));
+        verifyBatchMetadata(dataObject1BatchMessages.slice(1, 2));
+
+        // The next couple of ops in TurnBased mode should be in a batch have have batch metadata.
+        verifyBatchMetadata(dataObject1BatchMessages.slice(2, 4));
+
+        // The next couple of ops in Immediate mode should have been sent individually without batch metadata.
+        verifyBatchMetadata(dataObject1BatchMessages.slice(4, 5));
+        verifyBatchMetadata(dataObject1BatchMessages.slice(5, 6));
+
+        // The next couple of ops in TurnBased mode should be in a batch have have batch metadata.
+        verifyBatchMetadata(dataObject1BatchMessages.slice(6, 8));
+    });
+
+    it("can send ops alternatively with Immediate and TurnBased modes starting with TurnBased", async () => {
+        // Send couple of ops in TurnBased FlushMode. These ops should be batched together. No need to yield
+        // after sending these ops because setting FlushMode to Immediate will flush these ops.
+        dataObject1.context.containerRuntime.setFlushMode(FlushMode.TurnBased);
+        dataObject1map1.set("key1", "value1");
+        dataObject1map2.set("key2", "value2");
+
+        // Send couple of ops in Immediate FlushMode. These ops should not have batch metadata.
+        dataObject1.context.containerRuntime.setFlushMode(FlushMode.Immediate);
+        dataObject1map1.set("key3", "value3");
+        dataObject1map2.set("key4", "value4");
+
+        // Send couple of ops in TurnBased FlushMode. These ops should be batched together. No need to yield
+        // after sending these ops because setting FlushMode to Immediate will flush these ops.
+        dataObject1.context.containerRuntime.setFlushMode(FlushMode.TurnBased);
+        dataObject1map1.set("key5", "value5");
+        dataObject1map2.set("key6", "value6");
+
+        // Send couple of ops in Immediate FlushMode. These ops should not have batch metadata.
+        dataObject1.context.containerRuntime.setFlushMode(FlushMode.Immediate);
```

Comment:
@andre4i I found another bug with this test:
- The first FlushMode it sets is TurnBased. Since the default mode is the same, this is a no-op.
- The last FlushMode it sets is Immediate.
- After this, the container reconnects to transition from "read" to "write" mode for sending ops.
- When it now re-sends the first op, the FlushMode is the last one set which is Immediate. So, it does not batch the first two ops as we expect it to.

The fix here should be to add the default FlushMode to PendingStateManager's queue when the container runtime is created.

Reply:
Actually that fix won't work. The problem isn't specific to the initial flush mode but can happen any time. Basically, whenever we are replaying the states, we have to know what the flush mode was at that point in time because it could have been changed later.

Reply:
Is this the reason for https://dev.azure.com/fluidframework/public/_build/results?buildId=57352&view=logs&j=3dc8fd7e-4368-5a92-293e-d53cefc8c4b3 ?

Reply:
This link takes me to build output page. This test is failing in the CI run as well.

Reply:
I thought of a solution, let me try it.

Reply:
Added a fix - Track the last flush mode that was read from the pending queue which is basically the flush mode to be used for the subsequent ops. When replaying ops, set this flush mode to begin with.